Use spack-stack modulefiles

zmoon opened the pull request:
* [x] manually test the build on all platforms  
  e.g.
  ```
  rm -rf build && cmake -B build && cmake --build build -j 4 && mpirun -n 1 ./build/bin/nexus -h
  ```
  * [x] Hera Intel
  * [x] WCOSS
  * [x] Orion
  * [x] Gaea C5 (with the two adjustments)
  * [x] Derecho (note doesn't work on Casper)
* [x] revise the readme instructions

Closes #39

bbakernoaa committed:
Updating module files for building with current spack stack

bbakernoaa committed:
adding generic linux module build

bbakernoaa committed:
Merge remote-tracking branch 'origin/develop' into feature/modulefiles

bbakernoaa committed:
remove whitespace

bbakernoaa committed:
precommit issues

zmoon commented on CMakeLists.txt:
```diff
@@ -1,11 +1,20 @@
 cmake_minimum_required(VERSION 3.5)
 
-# Set the project name and version
+# set the project name and version
 project(NEXUS
         VERSION 0.2.0
         LANGUAGES Fortran)
 
+# set compiler
+if(NOT DEFINED ENV{FC})
+  message(FATAL_ERROR "CMAKE_Fortran_COMPILER is not defined")
+endif()
+
+set(CMAKE_Fortran_COMPILER $ENV{FC})
+set(CMAKE_C_COMPILER $ENV{CC})
+set(CMAKE_CXX_COMPILER $ENV{CXX})
+
```
This change is a reversion. CMake recognizes those env vars (`FC`, `CC`, etc.) automatically.

bbakernoaa committed:
remoe old files

bbakernoaa committed:
remove FC etc

zmoon commented on CMakeLists.txt:
```diff
@@ -1,11 +1,16 @@
 cmake_minimum_required(VERSION 3.5)
 
-# Set the project name and version
+# set the project name and version
```
```suggestion
# Set the project name and version
```

zmoon commented on CMakeLists.txt:
```diff
@@ -1,11 +1,16 @@
         VERSION 0.2.0
         LANGUAGES Fortran)
 
+# set compiler
+if(NOT DEFINED ENV{FC})
+  message(FATAL_ERROR "CMAKE_Fortran_COMPILER is not defined")
+endif()
```
It doesn't need to be fatal error. I was experimenting with these warning messages, but the behavior seems to be not always as expected, due to cmake's caching and such I guess.
```suggestion
# FC should be used to set compiler
if(NOT DEFINED ENV{FC})
  message(WARNING "FC is not set. You may want to use it to specify a specific compiler.")
else()
  if("$ENV{FC}" STREQUAL "")
    message(WARNING "FC is set to empty string. You may want to use it to specify a specific compiler.")
  else()
    message(STATUS "FC is set: '${FC}'")
  endif()
endif()
```

zmoon commented on modulefiles/ufs_common.lua:
```diff
@@ -0,0 +1,32 @@
+  {["netcdf-fortran"]  = "4.6.1"},
+  {["parallelio"]      = "2.5.10"},
+  {["esmf"]            = "8.6.0"},
+  {["fms"]             = "2024.01"},
+  {["bacio"]           = "2.4.1"},
+  {["crtm"]            = "2.4.0"},
+  {["g2"]              = "3.5.1"},
+  {["g2tmpl"]          = "1.13.0"},
+  {["ip"]              = "4.3.0"},
+  {["sp"]              = "2.5.0"},
+  {["w3emc"]           = "2.10.0"},
+  {["gftl-shared"]     = "1.6.1"},
+  {["mapl"]            = "2.40.3-esmf-8.6.0"},
+  {["scotch"]          = "7.0.4"},
```
I would guess we could remove these?

bbakernoaa replied:
We probably could. At least some of them. Definitely don’t need many of these like scotch and w3emc. These are for the wave model

bbakernoaa committed:
Update CMakeLists.txt

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

bbakernoaa committed:
Update CMakeLists.txt

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

bbakernoaa committed:
Update ufs_common.lua

zmoon committed:
Update readme with UFS module file info

and remove most of the Ubuntu instructions

bbakernoaa commented:
@zmoon Compiled fine on wcoss 

Did get a warning during the initial CMake step 

```bash
CMake Warning (dev) at src/CMakeLists.txt:16 (find_package):
  Policy CMP0074 is not set: find_package uses <PackageName>_ROOT variables.
  Run "cmake --help-policy CMP0074" for policy details.  Use the cmake_policy
  command to set the policy and suppress this warning.

  Environment variable ESMF_ROOT is set to:

    /apps/prod/hpc-stack/i-19.1.3.304__m-8.1.12__h-1.14.0__n-4.9.2__p-2.5.10__e-8.6.0pnetcdf/intel-19.1.3.304/cray-mpich-8.1.12/esmf/8.6.0

  For compatibility, CMake is ignoring the variable.
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Found ESMF: /apps/prod/hpc-stack/i-19.1.3.304__m-8.1.12__h-1.14.0__n-4.9.2__p-2.5.10__e-8.6.0pnetcdf/intel-19.1.3.304/cray-mpich-8.1.12/esmf/8.6.0/lib/libesmf.a (found version "8.6.0")
CMake Warning (dev) at src/CMakeLists.txt:17 (find_package):
  Policy CMP0074 is not set: find_package uses <PackageName>_ROOT variables.
  Run "cmake --help-policy CMP0074" for policy details.  Use the cmake_policy
  command to set the policy and suppress this warning.

  Environment variable NetCDF_ROOT is set to:

    /apps/prod/hpc-stack/i-19.1.3.304__m-8.1.12__h-1.14.0__n-4.9.2__p-2.5.10__e-8.6.0pnetcdf/intel-19.1.3.304/cray-mpich-8.1.12/netcdf/4.9.2

  For compatibility, CMake is ignoring the variable.
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Found NetCDF: /apps/prod/hpc-stack/i-19.1.3.304__m-8.1.12__h-1.14.0__n-4.9.2__p-2.5.10__e-8.6.0pnetcdf/intel-19.1.3.304/cray-mpich-8.1.12/netcdf/4.9.2/include (found version "4.9.2") found components: Fortran
-- FindNetCDF defines targets:
--   - NetCDF_VERSION [4.9.2]
--   - NetCDF_PARALLEL [TRUE]
--   - NetCDF_C_CONFIG_EXECUTABLE [/apps/prod/hpc-stack/i-19.1.3.304__m-8.1.12__h-1.14.0__n-4.9.2__p-2.5.10__e-8.6.0pnetcdf/intel-19.1.3.304/cray-mpich-8.1.12/netcdf/4.9.2/bin/nc-config]
--   - NetCDF::NetCDF_C [STATIC] [Root: /apps/prod/hpc-stack/i-19.1.3.304__m-8.1.12__h-1.14.0__n-4.9.2__p-2.5.10__e-8.6.0pnetcdf/intel-19.1.3.304/cray-mpich-8.1.12/netcdf/4.9.2] Lib: /apps/prod/hpc-stack/i-19.1.3.304__m-8.1.12__h-1.14.0__n-4.9.2__p-2.5.10__e-8.6.0pnetcdf/intel-19.1.3.304/cray-mpich-8.1.12/netcdf/4.9.2/lib/libnetcdf.a
--   - NetCDF_Fortran_CONFIG_EXECUTABLE [/apps/prod/hpc-stack/i-19.1.3.304__m-8.1.12__h-1.14.0__n-4.9.2__p-2.5.10__e-8.6.0pnetcdf/intel-19.1.3.304/cray-mpich-8.1.12/netcdf/4.9.2/bin/nf-config]
--   - NetCDF::NetCDF_Fortran [STATIC] [Root: /apps/prod/hpc-stack/i-19.1.3.304__m-8.1.12__h-1.14.0__n-4.9.2__p-2.5.10__e-8.6.0pnetcdf/intel-19.1.3.304/cray-mpich-8.1.12/netcdf/4.9.2] Lib: /apps/prod/hpc-stack/i-19.1.3.304__m-8.1.12__h-1.14.0__n-4.9.2__p-2.5.10__e-8.6.0pnetcdf/intel-19.1.3.304/cray-mpich-8.1.12/netcdf/4.9.2/lib/libnetcdff.a
CMake Warning (dev) at src/nuopc/CMakeLists.txt:16 (find_package):
  Policy CMP0074 is not set: find_package uses <PackageName>_ROOT variables.
  Run "cmake --help-policy CMP0074" for policy details.  Use the cmake_policy
  command to set the policy and suppress this warning.

  Environment variable ESMF_ROOT is set to:

    /apps/prod/hpc-stack/i-19.1.3.304__m-8.1.12__h-1.14.0__n-4.9.2__p-2.5.10__e-8.6.0pnetcdf/intel-19.1.3.304/cray-mpich-8.1.12/esmf/8.6.0

  For compatibility, CMake is ignoring the variable.
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at src/nuopc/CMakeLists.txt:17 (find_package):
  Policy CMP0074 is not set: find_package uses <PackageName>_ROOT variables.
  Run "cmake --help-policy CMP0074" for policy details.  Use the cmake_policy
  command to set the policy and suppress this warning.

  Environment variable NetCDF_ROOT is set to:

    /apps/prod/hpc-stack/i-19.1.3.304__m-8.1.12__h-1.14.0__n-4.9.2__p-2.5.10__e-8.6.0pnetcdf/intel-19.1.3.304/cray-mpich-8.1.12/netcdf/4.9.2

  For compatibility, CMake is ignoring the variable.
This warning is for project developers.  Use -Wno-dev to suppress it.
```

zmoon commented:
We could explicitly set the policy as it suggests. For example
```cmake
if(POLICY CMP0074)
  cmake_policy(SET CMP0074 OLD)
endif()
```
I don't see UWM doing something like that in their repo though.

zmoon committed:
Set CMake policy 0074

telling it to ignore _ROOT variables
since we have special find scripts for those

on WCOSS2, it was warning about these being set for NetCDF and ESMF

zmoon commented on modulefiles/ufs_gaea.intel.lua:
```diff
@@ -0,0 +1,33 @@
+
+prepend_path("MODULEPATH", "/ncrc/proj/epic/spack-stack/spack-stack-1.6.0/envs/fms-2024.01/install/modulefiles/Core")
+
+stack_intel_ver=os.getenv("stack_intel_ver") or "2023.1.0"
```
Tried C5 and C6 and it seems like only 2023.2.0 is available.
```suggestion
stack_intel_ver=os.getenv("stack_intel_ver") or "2023.2.0"
```

zmoon replied:
https://github.com/ufs-community/ufs-weather-model/commit/db1781a05dce1125cfe17f8324650674640f0a9e

zmoon commented on modulefiles/ufs_gaea.intel.lua:
```diff
@@ -0,0 +1,33 @@
+stack_intel_ver=os.getenv("stack_intel_ver") or "2023.1.0"
+load(pathJoin("stack-intel", stack_intel_ver))
+
+stack_cray_mpich_ver=os.getenv("stack_cray_mpich_ver") or "8.1.25"
```
Again seems like this is the only version.
```suggestion
stack_cray_mpich_ver=os.getenv("stack_cray_mpich_ver") or "8.1.28"
```

bbakernoaa committed:
Update modulefiles/ufs_gaea.intel.lua

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

bbakernoaa committed:
Update modulefiles/ufs_gaea.intel.lua

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

bbakernoaa committed:
Merge branch 'develop' into feature/modulefiles

zmoon commented:
@bbakernoaa does revised readme look okay to you?

zmoon commented on README.md:
```diff
@@ -52,58 +57,10 @@ Input data:
-- `ESMFMKFILE`
-- `OMPI_FC`
-- `CMAKE_Fortran_COMPILER`
+See https://github.com/zmoon/gha-esmf for examples of how to build ESMF,
```
Migrated
```suggestion
See https://github.com/noaa-oar-arl/gha-esmf for examples of how to build ESMF,
```

zmoon committed:
gha-esmf moved to ARL org

zmoon committed:
Try unsetting WCOSS vars instead

zmoon committed:
Print `$ENV{FC}`

note it seems CMake compiler discovery modifies this result
from command to full path

bbakernoaa commented:
whats left for this?

zmoon commented:
> whats left for this?

Were you able to try WCOSS from fresh again after the change? see if works and without warnings?

bbakernoaa commented:
compiled

zmoon commented:
> compiled

And you didn't get those warnings from earlier in this thread?

bbakernoaa commented:
No warning. Compiled just fine


Barry Baker

National Oceanic and Atmospheric Administration
Air Resources Laboratory
Physical Research Scientist
Chemical Modeling and Emissions Group Leader
NCWCP, R/ARL, Rm. 4204
5830 University Research Court
College Park, Maryland  20740
Phone: ‪(301) 683-1395‬



On Thu, Nov 14, 2024 at 6:27 PM Zachary Moon ***@***.***>
wrote:

> compiled
>
> And you didn't get those warnings from earlier in this thread?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/noaa-oar-arl/NEXUS/pull/56#issuecomment-2477615476>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFIUVNYLZJPYAN2COE4KBGL2AUWWLAVCNFSM6AAAAABQEMQYSOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDINZXGYYTKNBXGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

zmoon commented:
> No warning. Compiled just fine

Great, do you want to approve then (seems I can't since I opened this PR, even though it's from your fork)
